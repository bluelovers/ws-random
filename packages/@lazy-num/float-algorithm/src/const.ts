
export const MATH_POW_2_32 = Math.pow(2, 32);
