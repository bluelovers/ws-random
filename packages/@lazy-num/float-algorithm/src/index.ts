export * from './algorithm/number/mulberry32';
export * from './algorithm/number/splitmix32';
export * from './algorithm/int-list/sfc32';
export * from './algorithm/int-list/tyche';
