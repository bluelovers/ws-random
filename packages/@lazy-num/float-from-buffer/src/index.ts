import { FLOAT_ENTROPY_BYTES, MATH_POW_2_32 } from '@lazy-random/shared-lib';

/**
 * Given a buffer containing bytes of entropy, generate a double-precision
 * 64-bit float.
 *
 * @param {Buffer} buf a buffer of bytes
 * @returns {Number} a float
 *
 * @see https://github.com/fardog/node-random-lib/blob/master/index.js
 * @see http://stackoverflow.com/questions/15753019/floating-point-number-from-crypto-randombytes-in-javascript
 */
export function floatFromBuffer(buf: ArrayLike<number>, offset: number = 0)
{
	offset = Math.floor(offset);

	if (buf.length < (FLOAT_ENTROPY_BYTES + offset) || offset < 0)
	{
		throw new RangeError(
			`buffer must contain at least ${FLOAT_ENTROPY_BYTES}${offset > 0 ? ' +' + offset : ''} bytes of entropy`,
		)
	}

	return _floatFromBuffer(buf, offset)
}

export function _floatFromBuffer(buf: ArrayLike<number>, offset: number = 0)
{
	let position = Math.floor(offset);

	return (((((((
								buf[position++] % 32) / 32 +
							buf[position++]) / 256 +
						buf[position++]) / 256 +
					buf[position++]) / 256 +
				buf[position++]) / 256 +
			buf[position++]) / 256 +
		buf[position]) / 256
}

export function _floatFromBuffer2(buf: ArrayLike<number>, offset: number = 0)
{
	return readUInt32BE(buf, offset) / MATH_POW_2_32
}

export function readUInt32LE(buf: ArrayLike<number>, offset: number = 0)
{
	offset = offset >>> 0;

	return ((buf[offset]) |
			(buf[offset + 1] << 8) |
			(buf[offset + 2] << 16)) +
		(buf[offset + 3] * 0x1000000)
}

export function readUInt32BE(buf: ArrayLike<number>, offset: number = 0)
{
	offset = offset >>> 0;

	return (buf[offset] * 0x1000000) +
		((buf[offset + 1] << 16) |
			(buf[offset + 2] << 8) |
			buf[offset + 3])
}

export default floatFromBuffer
