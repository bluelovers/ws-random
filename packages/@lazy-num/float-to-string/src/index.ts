export function assertFractionDigits(fractionDigits?: number): asserts fractionDigits is number
{
	if (!Number.isFinite(fractionDigits) || !Number.isInteger(fractionDigits) || fractionDigits === 0)
	{
		throw new TypeError(`Invalid fractionDigits: ${fractionDigits}`)
	}
}

export function splitFloatNumberToString(float: string | number): [string, string?]
{
	return String(float).split('.') as any
}

export function getFractionDigitsString(float: string | number)
{
	return splitFloatNumberToString(float)[1]
}

export function splitFloatNumber(n: number): [number, number]
{
	let int = Math.floor(n);
	let float = n - int;

	return [int, float]
}

export function joinFloatNumber(int: number | string, float?: string)
{
	return String(int) + (float?.length ? '.' + float : '');
}

export function floatToString(n: number, fractionDigits?: number)
{
	let int: string | number;
	let s: string;

	if (typeof fractionDigits === 'number')
	{
		assertFractionDigits(fractionDigits);

		([int, s] = splitFloatNumberToString(n.toFixed(fractionDigits)));
	}
	else
	{
		let float;
		([int, float] = splitFloatNumber(n));
		s = getFractionDigitsString(float);
	}

	return joinFloatNumber(int, s);
}

export default floatToString
