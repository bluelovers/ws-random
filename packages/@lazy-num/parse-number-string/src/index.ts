
const reInt = /^[+-]?\d+$/;
const reFloatOnly = /^[+-]?(?:\d+)?\.\d+$/;
const reFloat = new RegExp(reInt.source + '|' + reFloatOnly.source);

export function isIntString<T extends `${number}` | string = `${number}`>(input: unknown): input is T
{
	return (typeof input === 'string' && reInt.test(input))
}

export function isFloatOnlyString<T extends `${number}` | string = `${number}`>(input: unknown | T): input is T
{
	return (typeof input === 'string' && reFloatOnly.test(input))
}

export function isFloatString<T extends `${number}` | string = `${number}`>(input: unknown | T): input is T
{
	return (typeof input === 'string' && reFloat.test(input))
}

export type IParseNumberFn = (<T extends `${number}` | string = `${number}`>(input: unknown | T) => input is T) | ((input: unknown) => input is`${number}`)

export function _parseNumberString<T extends number>(validFn: IParseNumberFn, input: unknown | T | `${T}`): T
{
	if (validFn(input))
	{
		return Number(input) as T
	}

	if (typeof input !== 'number')
	{
		throw new TypeError(`Invalid value: ${input}`)
	}

	return input as T
}

export function parseIntString<T extends number>(input: T): T
export function parseIntString<T extends number>(input: `${T}`): T
export function parseIntString<T extends number>(input: unknown | T | `${T}`): T
export function parseIntString<T extends number>(input: unknown | T | `${T}`): T
{
	return _parseNumberString(isIntString, input)
}

export function parseFloatOnlyString<T extends number>(input: T): T
export function parseFloatOnlyString<T extends number>(input: `${T}`): T
export function parseFloatOnlyString<T extends number>(input: unknown | T | `${T}`): T
export function parseFloatOnlyString<T extends number>(input: unknown | T | `${T}`): T
{
	return _parseNumberString(isFloatOnlyString, input)
}

export function parseFloatString<T extends number>(input: T): T
export function parseFloatString<T extends number>(input: `${T}`): T
export function parseFloatString<T extends number>(input: unknown | T | `${T}`): T
export function parseFloatString<T extends number>(input: unknown | T | `${T}`): T
{
	return _parseNumberString(isFloatString, input as any)
}

export default parseFloatString
