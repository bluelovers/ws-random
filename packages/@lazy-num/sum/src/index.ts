import { fixZero } from 'num-is-zero';

/**
 * (1 + 2 + 3 +...+N)
 *
 * @see http://emn178.pixnet.net/blog/post/92132837-%E8%A8%88%E7%AE%971%E5%88%B0n%E7%B8%BD%E5%92%8C%281-%2B-2-%2B-3-%2B...%2Bn%29
 */
export function sum_1_to_n(n: number)
{
	return n * (n + 1) / 2;
}

export function num_array_sum(na: number[])
{
	return fixZero(na.reduce((a, b) => a + b))
}
