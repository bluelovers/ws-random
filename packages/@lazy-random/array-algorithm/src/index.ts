
export * from './swapAlgorithm';
export * from './array_rebase';
