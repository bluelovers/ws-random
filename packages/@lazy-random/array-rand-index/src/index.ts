import { _MathRandom } from '@lazy-random/original-math-random';

export function arrayRandIndexByLength(len: number, ...argv: any[])
{
	return Math.floor(_MathRandom() * len)
}

export function arrayRandIndex(array: any[], ...argv: any[])
{
	return arrayRandIndexByLength(array.length)
}

export default arrayRandIndex
