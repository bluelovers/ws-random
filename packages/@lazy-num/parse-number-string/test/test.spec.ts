//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { basename, extname } from 'path';
import { parseFloatString, parseFloatOnlyString, parseIntString } from '../src/index';

beforeAll(async () =>
{

});

describe(basename(__filename, extname(__filename)), () =>
{

	test.skip(`dummy`, () => {});

	test.each([

		'0',
		'0.0',
		'0.1',
		'.1',

		'+0',
		'+0.0',
		'+0.1',
		'+.1',

		'-0',
		'-0.0',
		'-0.1',
		'-.1',

	])(`%j`, (input) =>
	{
		let actual = parseFloatString(input);
		let expected;

		expect(typeof actual).toStrictEqual('number');

		let b;
		let c;

		try
		{
			b = parseFloatOnlyString(input);
		}
		catch (e)
		{

		}

		try
		{
			b = parseFloatOnlyString(input);
		}
		catch (e)
		{
			c = parseIntString(input)
		}

		expect({
			parseFloatString: actual,
			parseFloatOnlyString: b,
			parseIntString: c,
		}).toMatchSnapshot();

	});

})
