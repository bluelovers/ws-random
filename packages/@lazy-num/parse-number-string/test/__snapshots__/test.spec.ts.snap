// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test.spec "+.1" 1`] = `
{
  "parseFloatOnlyString": 0.1,
  "parseFloatString": 0.1,
  "parseIntString": undefined,
}
`;

exports[`test.spec "+0" 1`] = `
{
  "parseFloatOnlyString": undefined,
  "parseFloatString": 0,
  "parseIntString": 0,
}
`;

exports[`test.spec "+0.0" 1`] = `
{
  "parseFloatOnlyString": 0,
  "parseFloatString": 0,
  "parseIntString": undefined,
}
`;

exports[`test.spec "+0.1" 1`] = `
{
  "parseFloatOnlyString": 0.1,
  "parseFloatString": 0.1,
  "parseIntString": undefined,
}
`;

exports[`test.spec ".1" 1`] = `
{
  "parseFloatOnlyString": 0.1,
  "parseFloatString": 0.1,
  "parseIntString": undefined,
}
`;

exports[`test.spec "-.1" 1`] = `
{
  "parseFloatOnlyString": -0.1,
  "parseFloatString": -0.1,
  "parseIntString": undefined,
}
`;

exports[`test.spec "-0" 1`] = `
{
  "parseFloatOnlyString": undefined,
  "parseFloatString": -0,
  "parseIntString": -0,
}
`;

exports[`test.spec "-0.0" 1`] = `
{
  "parseFloatOnlyString": -0,
  "parseFloatString": -0,
  "parseIntString": undefined,
}
`;

exports[`test.spec "-0.1" 1`] = `
{
  "parseFloatOnlyString": -0.1,
  "parseFloatString": -0.1,
  "parseIntString": undefined,
}
`;

exports[`test.spec "0" 1`] = `
{
  "parseFloatOnlyString": undefined,
  "parseFloatString": 0,
  "parseIntString": 0,
}
`;

exports[`test.spec "0.0" 1`] = `
{
  "parseFloatOnlyString": 0,
  "parseFloatString": 0,
  "parseIntString": undefined,
}
`;

exports[`test.spec "0.1" 1`] = `
{
  "parseFloatOnlyString": 0.1,
  "parseFloatString": 0.1,
  "parseIntString": undefined,
}
`;
